beancount/core/amount.py

```diff
@@ -48,6 +48,22 @@ def __add__(self,other):
     def __sub__(self,other):
         return sub(self,other)
 
+    def __mul__(self,other):
+        try:
+            other = Decimal(other)
+        except ValueError:
+            pass
```

Comment:
Why the attempted implicit conversion? The right way to implement this is:
```python
def __mul__(self, other):
    if not isinstance(other, Decimal):
        return NotImplemented
    return mul(self, other)
```
and similarly for `__truediv__`.

Probably `__add__` and `__sub__` should be in the form:
```python
def __add__(self, other):
    if not isinstance(other, Amount):
        return NotImplemented
    return add(self, other)
```

And at this point most of the asserts in the `add()` and `mul()` functions is redundant and maybe open coding them in the method implementations could be worth for performance.

Reply:
The reason for the converting to Decimal is to support Numbers /  Ints (and I guess floats come from the ride). If I run:

select 4*value(position)

It will error without the conversion because 'Number is not a Decimal'. If you prefer, I could explicitly check for type Number or Decimal (and convert Numbers), but currently the behavior is to try to make the input a Decimal and if that fails it'll pass the original type which will then return a helpful error. If you try to multiply a Position by an Amount, you'll get:

 unsupported operand type(s) for /: 'Position' and 'Amount'

which seems like reasonable behavior. Another option would be to edit the add function itself to support more than just Decimal inputs, but since that function is (potentially) used elsewhere, I thought it best not to touch it. Let me know what you think is best.

Reply:
AFAIK, the BQL parser only produces `Decimal` instances.

Reply:
Correction, the BQL parser also emits `int`s. The check should be `isinstance(other, (Decimal, int))` and the implementation should support that: multiplying a `Decimal` by an `int` is supported and well defined.

Reply:
Sounds good. Just to clarify, do you think its better in the __add__ function to check the instance and if it is an int, convert to a Decimal? Or check just check and pass Decimal or Int to the add function and then edit the add function to work with type int in addition to Decimal?

Reply:
That's a good point. My latest (squashed) commit tries to address this. Instead of handling it in the __mul__ function, I put it in the EvalMul function to check the types and respond accordingly, which I think might be a little cleaner.

Reply:
Done.